### ISSUE TITLE
A Comprehensive guide to Fine-tuning Deep Learning Models in Keras (Part II)

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9669]
A Comprehensive guide to Fine-tuning Deep Learning Models in Keras (Part II)<br>
<br>
https://ift.tt/REOnw36<br>
<br>
<br>
<br>

<div><div><div><div><section><p>This is Part II of a 2 part series that cover fine-tuning deep learning models in Keras. <a href="https://flyyufelix.github.io/2016/10/03/fine-tuning-in-keras-part1.html" target="_blank">Part I</a> states the motivation and rationale behind fine-tuning and gives a brief introduction on the common practices and techniques. This post will give a detailed step-by-step guide on how to go about implementing fine-tuning on popular models <em>VGG</em>, <em>Inception</em>, and <em>ResNet</em> in Keras.</p>
<h2>Why do we pick Keras?</h2>
<p><a href="https://keras.io/" target="_blank">Keras</a> is a simple to use neural network library built on top of Theano or TensorFlow that allows developers to prototype ideas very quickly. Unless you are doing some cutting-edge research that involves customizing a completely novel neural architecture with different activation mechanism, Keras provides all the building blocks you need to build reasonably sophisticated neural networks.</p>
<p>It also comes with a great documentation and tons of online resources.</p>
<h2>Note on Hardware</h2>
<p>I would strongly suggest getting a GPU to do the heavy computation involved in Covnet training. The speed difference is very substantial. We are talking about a matter of hours with a GPU versus a matter of days with a CPU.</p>
<p>I would recommend GTX 980Ti or a slightly expensive GTX 1080 which cost around $600 bucks.</p>
<h2>Fine-tuning in Keras</h2>
<p>I have implemented starter scripts for fine-tuning convnets in Keras. The scripts are hosted in <a href="https://github.com/flyyufelix/cnn_finetune" target="_blank">this github page</a>. Implementations of VGG16, VGG19, GoogLeNet, Inception-V3, and ResNet50 are included. With that, you can customize the scripts for your own fine-tuning task.</p>
<p>Below is a detailed walkthrough of how to fine-tune <strong>VGG16</strong> and <strong>Inception-V3</strong> models using the scripts.</p>
<p><strong>Fine-tune VGG16.</strong> VGG16 is a 16-layer Covnet used by the <a href="http://www.robots.ox.ac.uk/~vgg/research/very_deep/" target="_blank">Visual Geometry Group</a> (VGG) at Oxford University in the 2014 ILSVRC (ImageNet) competition. The model achieves a 7.5% top 5 error rate on the validation set, which is a result that earned them a second place finish in the competition.</p>
<p>Schematic Diagram of VGG16 Model:</p>
<p><img alt="VGG16 Model" src="https://flyyufelix.github.io/img/vgg16.png"></p>
<p>The script for fine-tuning VGG16 can be found in <a href="https://github.com/flyyufelix/cnn_finetune/blob/master/vgg16.py" target="_blank">vgg16.py</a>. The first part of the <code>vgg_std16_model</code> function is the model schema for VGG16. After defining the fully connected layer, we load the ImageNet pre-trained weight to the model by the following line:</p>
<figure><pre><code><span>model</span><span>.</span><span>load_weights</span><span>(</span><span>'cache/vgg16_weights.h5'</span><span>)</span></code></pre></figure><p>For fine-tuning purpose, we truncate the original softmax layer and replace it with our own by the following snippet:</p>
<figure><pre><code><span>model</span><span>.</span><span>layers</span><span>.</span><span>pop</span><span>()</span>
<span>model</span><span>.</span><span>outputs</span> <span>=</span> <span>[</span><span>model</span><span>.</span><span>layers</span><span>[</span><span>-</span><span>1</span><span>].</span><span>output</span><span>]</span>
<span>model</span><span>.</span><span>layers</span><span>[</span><span>-</span><span>1</span><span>].</span><span>outbound_nodes</span> <span>=</span> <span>[]</span>
<span>model</span><span>.</span><span>add</span><span>(</span><span>Dense</span><span>(</span><span>num_class</span><span>,</span> <span>activation</span><span>=</span><span>'softmax'</span><span>))</span></code></pre></figure><p>Where the <code>num_class</code> variable in the last line represents the number of class labels for our classification task.</p>
<p>Sometimes, we want to freeze the weight for the first few layers so that they remain intact throughout the fine-tuning process. Say we want to freeze the weights for the first 10 layers. This can be done by the following lines:</p>
<figure><pre><code><span>for</span> <span>layer</span> <span>in</span> <span>model</span><span>.</span><span>layers</span><span>[:</span><span>10</span><span>]:</span>
    <span>layer</span><span>.</span><span>trainable</span> <span>=</span> <span>False</span></code></pre></figure><p>We then fine-tune the model by minimizing the cross entropy loss function using stochastic gradient descent (sgd) algorithm. Notice that we use an initial learning rate of 0.001, which is smaller than the learning rate for training scratch model (usually 0.01).</p>
<figure><pre><code><span>sgd</span> <span>=</span> <span>SGD</span><span>(</span><span>lr</span><span>=</span><span>1e-3</span><span>,</span> <span>decay</span><span>=</span><span>1e-6</span><span>,</span> <span>momentum</span><span>=</span><span>0.9</span><span>,</span> <span>nesterov</span><span>=</span><span>True</span><span>)</span>
<span>model</span><span>.</span><span>compile</span><span>(</span><span>optimizer</span><span>=</span><span>sgd</span><span>,</span> <span>loss</span><span>=</span><span>'categorical_crossentropy'</span><span>,</span> <span>metrics</span><span>=</span><span>[</span><span>'accuracy'</span><span>])</span>
<span>model</span> <span>=</span> <span>vgg_std16_model</span><span>(</span><span>img_rows</span><span>,</span> <span>img_cols</span><span>,</span> <span>channel</span><span>,</span> <span>num_class</span><span>)</span></code></pre></figure><p>Where img_rows, img_cols, and channel define the dimension of the input. For colored image with resolution 224x224, img_rows = img_cols = 224, channel = 3.</p>
<p>Next, we load our dataset, split it into training and testing sets, and start fine-tuning the model:</p>
<figure><pre><code><span>X_train</span><span>,</span> <span>X_valid</span><span>,</span> <span>Y_train</span><span>,</span> <span>Y_valid</span> <span>=</span> <span>load_data</span><span>()</span>

<span>model</span><span>.</span><span>fit</span><span>(</span><span>train_data</span><span>,</span> <span>test_data</span><span>,</span>
          <span>batch_size</span><span>=</span><span>batch_size</span><span>,</span>
          <span>nb_epoch</span><span>=</span><span>nb_epoch</span><span>,</span>
          <span>shuffle</span><span>=</span><span>True</span><span>,</span>
          <span>verbose</span><span>=</span><span>1</span><span>,</span>
          <span>validation_data</span><span>=</span><span>(</span><span>X_valid</span><span>,</span> <span>Y_valid</span><span>),</span>
          <span>)</span></code></pre></figure><p>The fine-tuning process will take a while, depending on your hardware. After it is done, we use the model the make prediction on the validation set and return the score for the cross entropy loss:</p>
<figure><pre><code><span>predictions_valid</span> <span>=</span> <span>model</span><span>.</span><span>predict</span><span>(</span><span>X_valid</span><span>,</span> <span>batch_size</span><span>=</span><span>batch_size</span><span>,</span> <span>verbose</span><span>=</span><span>1</span><span>)</span>
<span>score</span> <span>=</span> <span>log_loss</span><span>(</span><span>Y_valid</span><span>,</span> <span>predictions_valid</span><span>)</span></code></pre></figure><p><strong>Fine-tune Inception-V3.</strong> Inception-V3 achieved the second place in the 2015 ImageNet competition with a 5.6% top 5 error rate on the validation set. The model is characterized by the usage of the <a href="https://www.youtube.com/watch?v=VxhSouuSZDY" target="_blank">Inception Module</a>, which is a concatenation of features maps generated by kernels of varying dimensions.</p>
<p>Schematic Diagram of the 27-layer Inception-V1 Model (Idea similar to that of V3):</p>
<p><img alt="Inception-V1 Model" src="https://flyyufelix.github.io/img/inception.png"></p>
<p>The code for fine-tuning Inception-V3 can be found in <a href="https://github.com/flyyufelix/cnn_finetune/blob/master/inception_v3.py" target="_blank">inception_v3.py</a>. The process is mostly similar to that of VGG16, with one subtle difference. Inception-V3 does not use Keras’ <a href="https://keras.io/models/sequential/" target="_blank">Sequential Model</a> due to branch merging (for the inception module), hence we cannot simply use <code>model.pop()</code> to truncate the top layer.</p>
<p>Instead, after we create the model and load it up with the ImageNet weight, we perform the equivalent of top layer truncation by defining another fully connected sofmax (<code>x_newfc</code>) on top of the last inception module (x). This is done using the following snippet:</p>
<figure><pre><code><span># Last Inception Module </span><span>x</span> <span>=</span> <span>merge</span><span>([</span><span>branch1x1</span><span>,</span> <span>branch3x3</span><span>,</span> <span>branch3x3dbl</span><span>,</span> <span>branch_pool</span><span>],</span>
                  <span>mode</span><span>=</span><span>'concat'</span><span>,</span> <span>concat_axis</span><span>=</span><span>channel_axis</span><span>,</span>
                  <span>name</span><span>=</span><span>'mixed'</span> <span>+</span> <span>str</span><span>(</span><span>9</span> <span>+</span> <span>i</span><span>))</span>

<span># Fully Connected Softmax Layer
</span><span>x_fc</span> <span>=</span> <span>AveragePooling2D</span><span>((</span><span>8</span><span>,</span> <span>8</span><span>),</span> <span>strides</span><span>=</span><span>(</span><span>8</span><span>,</span> <span>8</span><span>),</span> <span>name</span><span>=</span><span>'avg_pool'</span><span>)(</span><span>x</span><span>)</span>
<span>x_fc</span> <span>=</span> <span>Flatten</span><span>(</span><span>name</span><span>=</span><span>'flatten'</span><span>)(</span><span>x_fc</span><span>)</span>
<span>x_fc</span> <span>=</span> <span>Dense</span><span>(</span><span>1000</span><span>,</span> <span>activation</span><span>=</span><span>'softmax'</span><span>,</span> <span>name</span><span>=</span><span>'predictions'</span><span>)(</span><span>x_fc</span><span>)</span>

<span># Create model
</span><span>model</span> <span>=</span> <span>Model</span><span>(</span><span>img_input</span><span>,</span> <span>x_fc</span><span>)</span>

<span># Load ImageNet pre-trained data
</span><span>model</span><span>.</span><span>load_weights</span><span>(</span><span>'cache/inception_v3_weights_th_dim_ordering_th_kernels.h5'</span><span>)</span>

<span># Truncate and replace softmax layer for transfer learning
# Cannot use model.layers.pop() since model is not of Sequential() type
# The method below works since pre-trained weights are stored in layers but not in the model
</span><span>x_newfc</span> <span>=</span> <span>AveragePooling2D</span><span>((</span><span>8</span><span>,</span> <span>8</span><span>),</span> <span>strides</span><span>=</span><span>(</span><span>8</span><span>,</span> <span>8</span><span>),</span> <span>name</span><span>=</span><span>'avg_pool'</span><span>)(</span><span>x</span><span>)</span>
<span>x_newfc</span> <span>=</span> <span>Flatten</span><span>(</span><span>name</span><span>=</span><span>'flatten'</span><span>)(</span><span>x_newfc</span><span>)</span>

<span># Create another model with our customized softmax
</span><span>model</span> <span>=</span> <span>Model</span><span>(</span><span>img_input</span><span>,</span> <span>x_newfc</span><span>)</span>

<span># Learning rate is changed to 0.001
</span><span>sgd</span> <span>=</span> <span>SGD</span><span>(</span><span>lr</span><span>=</span><span>1e-3</span><span>,</span> <span>decay</span><span>=</span><span>1e-6</span><span>,</span> <span>momentum</span><span>=</span><span>0.9</span><span>,</span> <span>nesterov</span><span>=</span><span>True</span><span>)</span>
<span>model</span><span>.</span><span>compile</span><span>(</span><span>optimizer</span><span>=</span><span>sgd</span><span>,</span> <span>loss</span><span>=</span><span>'categorical_crossentropy'</span><span>,</span> <span>metrics</span><span>=</span><span>[</span><span>'accuracy'</span><span>])</span>
<span>model</span> <span>=</span> <span>inception_v3_model</span><span>(</span><span>img_rows</span><span>,</span> <span>img_cols</span><span>,</span> <span>channel</span><span>,</span> <span>num_class</span><span>)</span></code></pre></figure><p>That’s it for Inception-V3. Starting script for other models such as <strong>VGG19</strong>, <strong>GoogleLeNet</strong>, and <strong>ResNet</strong> can be found <a href="https://github.com/flyyufelix/cnn_finetune" target="_blank">here</a>.</p>
<h2>Fine-tuned Networks in Action</h2>
<p>If you are a Deep Learning or Computer Vision practitioner, most likely you have already tried fine-tuning pre-trained network for your own classification problem before.</p>
<p>To me, I came across this interesting <a href="https://www.kaggle.com/c/state-farm-distracted-driver-detection" target="_blank">Kaggle Competition</a> which requires candidates to identify distracted drivers through analyzing in-car camera images. This is a good opportunity for me to try out fine-tuning in Keras.</p>
<p>Following the fine-tuning methods listed above, together with data preprocessing/augmentation and model ensembling our team managed to achieved top 4% in the competition. Detailed account of our method and lessons learned are captured in <a href="https://flyyufelix.github.io/2016/10/11/kaggle-statefarm.html" target="_blank">this post</a>.</p>
<p>If you have any questions or thoughts feel free to leave a comment below.</p>
<p>You can also follow me on Twitter at @flyyufelix. <br></p>
<div class="feedlyNoScript">Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a>
</div>
</section></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via GitHub Pages | Websites for you and your projects, hosted directly from your GitHub repository. Just edit, push, and your changes are live. <br>
<br>
July 23, 2023 at 06:29PM
